src/misc/ethernet_slirp.cpp

```diff
@@ -0,0 +1,267 @@
+	return conn->Poll_Get_Slirp_Revents(idx);
+}
+
+void slirp_register_poll_fd(int fd, void *opaque)
+{
+	(void)fd; (void)opaque;
+	/* TODO: find way to preserve these across polls_clear */
+	/*polls_add(fd, POLLIN | POLLOUT);*/
+}
+
+void slirp_unregister_poll_fd(int fd, void *opaque)
+{
+	(void)fd; (void)opaque;
+	/*polls_remove(fd);*/
+}
+
+void slirp_notify(void *opaque)
+{
+	(void)opaque;
+	return;
+}
+
+SlirpEthernetConnection::SlirpEthernetConnection(void)
+      : EthernetConnection()
+{
+	slirp_callbacks.send_packet = slirp_send_packet;
+	slirp_callbacks.guest_error = slirp_guest_error;
+	slirp_callbacks.clock_get_ns = slirp_clock_get_ns;
+	slirp_callbacks.timer_new = slirp_timer_new;
+	slirp_callbacks.timer_free = slirp_timer_free;
+	slirp_callbacks.timer_mod = slirp_timer_mod;
+	slirp_callbacks.register_poll_fd = slirp_register_poll_fd;
+	slirp_callbacks.unregister_poll_fd = slirp_unregister_poll_fd;
+	slirp_callbacks.notify = slirp_notify;
+}
+
+SlirpEthernetConnection::~SlirpEthernetConnection(void)
+{
+	if(slirp) slirp_cleanup(slirp);
+}
+
+bool SlirpEthernetConnection::Initialize(Section* dosbox_config)
+{
+	(void)dosbox_config; /* Ignore for now */
+
+	/* Config */
+	config.version = 1;
+	config.restricted = 0; /* Allow access to host */
+	config.disable_host_loopback = 0; /* Allow access to 127.0.0.1 */
+	config.if_mtu = 0; /* IF_MTU_DEFAULT */
+	config.if_mru = 0; /* IF_MRU_DEFAULT */
+	config.enable_emu = 0; /* Buggy, don't use */
+	/* IPv4 */
+	config.in_enabled = 1;
+	inet_pton(AF_INET, "10.0.2.0", &config.vnetwork);
+	inet_pton(AF_INET, "255.255.255.0", &config.vnetmask);
+	inet_pton(AF_INET, "10.0.2.2", &config.vhost);
+	inet_pton(AF_INET, "10.0.2.3", &config.vnameserver);
+	inet_pton(AF_INET, "10.0.2.15", &config.vdhcp_start);
+	/* IPv6 code is left here as reference but disabled as no DOS-era
+	 * software supports it and might get confused by it */
+	config.in6_enabled = 0;
+	inet_pton(AF_INET6, "fec0::", &config.vprefix_addr6);
+	config.vprefix_len = 64;
+	inet_pton(AF_INET6, "fec0::2", &config.vhost6);
+	inet_pton(AF_INET6, "fec0::3", &config.vnameserver6);
+	/* DHCPv4, BOOTP, TFTP */
+	config.vhostname = "DOSBox-X";
+	config.vdnssearch = NULL;
+	config.vdomainname = NULL;
+	config.tftp_server_name = NULL;
+	config.tftp_path = NULL;
+	config.bootfile = NULL;
+
+	slirp = slirp_new(&config, &slirp_callbacks, this);
+	if(slirp)
+	{
+		LOG_MSG("SLIRP successfully initialized");
+		return true;
+	}
+	else
+	{
+		/* TODO: better error message? */
+		LOG_MSG("SLIRP failed to initialize");
+		return false;
+	}
+}
+
+void SlirpEthernetConnection::SendPacket(uint8_t* packet, int len)
+{
+	slirp_input(slirp, packet, len);
+}
+
+void SlirpEthernetConnection::GetPackets(std::function<void(uint8_t*, int)> callback)
+{
+	get_packet_callback = callback;
+	uint32_t timeout = 0;
+	Polls_Clear();
+	slirp_pollfds_fill(slirp, &timeout, slirp_add_poll, this);
+	/* TODO: replace poll with select */
+	int ret = poll(polls, 256, timeout);
```

Comment:
You mentioned making using (or needing)  `select` for Windows.
Is this still under investigation?